tests/bgp/test_bgp_gr_helper.py

```diff
@@ -156,44 +52,88 @@ def _get_learned_bgp_routes_from_neighbor(duthost, bgp_neighbor):
             if prefix not in rib:
-                pytest.fail("Route to prefix %s doesn't exist after graceful restart." % prefix)
+                pytest.fail("Route to prefix %s doesn't exist during graceful restart." % prefix)
             nexthop_expected = nexthops[0]
             bgp_neighbor_expected = nexthop_expected["peerId"]
             for nexthop in rib[prefix]:
                 if nexthop["peerId"] == bgp_neighbor_expected:
                     if nexthop.get("stale", False) is False:
-                        pytest.fail(
-                            "Route to prefix %s should be stale after graceful restart, before: %s, after: %s" % (prefix, nexthop_expected, rib[prefix])
-                        )
+                        logging.error("Rib route entry to prefix %s:\n%s\n", prefix, json.dumps(rib[prefix]))
+                        pytest.fail("Route to prefix %s should be stale during graceful restart." % prefix)
                     break
             else:
-                pytest.fail(
-                    "Route to prefix doesn't originate from BGP neighbor %s, before: %s, after: %s" % (bgp_neighbor_expected, nexthop_expected, rib[prefix])
-                )
+                logging.error("Rib route entry to prefix %s:\n%s\n", prefix, json.dumps(rib[prefix]))
+                pytest.fail("Route to prefix doesn't originate from BGP neighbor %s." % bgp_neighbor_expected)
+
+    def _verify_prefix_counters_from_neighbor_after_graceful_restart(duthost, bgp_neighbors):
+        """Verify routes from neighbor are relearned and out of stale after graceful restart."""
+        for bgp_neighbor in bgp_neighbors:
+            for namespace in duthost.get_frontend_asic_namespace_list():
+                counters = _get_prefix_counters(duthost, bgp_neighbor, namespace)
+                logging.debug("Prefix counters for bgp neighbor %s in namespace %s:\n%s\n", bgp_neighbor, namespace, json.dumps(counters))
+                if not (counters["ribTableWalkCounters"]["Stale"] == 0 and counters["ribTableWalkCounters"]["Valid"] > 0):
+                    return False
+        return True
 
     duthost = duthosts[rand_one_dut_hostname]
 
     config_facts = duthost.config_facts(host=duthost.hostname, source="running")['ansible_facts']
     bgp_neighbors = config_facts.get('BGP_NEIGHBOR', {})
+    portchannels = config_facts.get('PORTCHANNEL', {})
     dev_nbrs = config_facts.get('DEVICE_NEIGHBOR', {})
+    configurations = tbinfo['topo']['properties']['configuration_properties']
+    exabgp_ips = [configurations['common']['nhipv4'], configurations['common']['nhipv6']]
+    exabgp_sessions = ['exabgp_v4', 'exabgp_v6']
+
+    # select neighbor to test
+    if "backend" in tbinfo["topo"]["name"]:
+        # for storage backend topologies, randomly choose a neighbor
+        test_interface = random.sample([k for k, v in dev_nbrs.items() if not v['name'].startswith("Server")], 1)[0]
```

Comment:
wonder if this could be more general, let's try to get the default route, and it default route is not available, then we choose start random neighbor?

Reply:
Sure, will update.